pkg/kubelet/status/state/checkpoint.go

```diff
@@ -18,48 +18,62 @@ package state
 
 import (
 	"encoding/json"
-
 	"k8s.io/api/core/v1"
 	"k8s.io/kubernetes/pkg/kubelet/checkpointmanager"
 	"k8s.io/kubernetes/pkg/kubelet/checkpointmanager/checksum"
 )
 
-var _ checkpointmanager.Checkpoint = &PodResourceAllocationCheckpoint{}
+var _ checkpointmanager.Checkpoint = &Checkpoint{}
 
-// PodResourceAllocationCheckpoint is used to store resources allocated to a pod in checkpoint
-type PodResourceAllocationCheckpoint struct {
+type PodResourceAllocationInfo struct {
 	AllocationEntries   map[string]map[string]v1.ResourceList `json:"allocationEntries,omitempty"`
 	ResizeStatusEntries map[string]v1.PodResizeStatus         `json:"resizeStatusEntries,omitempty"`
-	Checksum            checksum.Checksum                     `json:"checksum"`
 }
 
-// NewPodResourceAllocationCheckpoint returns an instance of Checkpoint
-func NewPodResourceAllocationCheckpoint() *PodResourceAllocationCheckpoint {
-	//lint:ignore unexported-type-in-api user-facing error message
-	return &PodResourceAllocationCheckpoint{
-		AllocationEntries:   make(map[string]map[string]v1.ResourceList),
-		ResizeStatusEntries: make(map[string]v1.PodResizeStatus),
+// Checkpoint represents a structure to store pod resource allocation checkpoint data
+type Checkpoint struct {
+	// Data is a JSON serialized checkpoint data
+	Data string `json:"data"`
+	// Checksum is a checksum of Data
+	Checksum checksum.Checksum `json:"checksum"`
+}
+
+// NewCheckpoint creates a new checkpoint from a list of claim info states
+func NewCheckpoint(data *PodResourceAllocationInfo) (*Checkpoint, error) {
+
+	praData, err := json.Marshal(data)
+	if err != nil {
+		return nil, err
+	}
+
+	cp := &Checkpoint{
+		Data: string(praData),
 	}
+	cp.Checksum = checksum.New(cp.Data)
+	return cp, nil
+}
+
+func (cp *Checkpoint) MarshalCheckpoint() ([]byte, error) {
+	return json.Marshal(cp)
 }
 
-// MarshalCheckpoint returns marshalled checkpoint
-func (prc *PodResourceAllocationCheckpoint) MarshalCheckpoint() ([]byte, error) {
-	// make sure checksum wasn't set before so it doesn't affect output checksum
-	prc.Checksum = 0
-	prc.Checksum = checksum.New(prc)
-	return json.Marshal(*prc)
+// UnmarshalCheckpoint unmarshals checkpoint from JSON
+func (cp *Checkpoint) UnmarshalCheckpoint(blob []byte) error {
+	return json.Unmarshal(blob, cp)
```

Comment:
Should UnmarshalCheckpoint verify the checksum?

Reply:
@bart0sh Per my understanding, we shouldn't verify the checksum in UnmarshalCheckpoint function. Since checkpoint interface provides separate function to verify checksum, we should keep it simple
```golang
// Checkpoint provides the process checkpoint data
type Checkpoint interface {
	MarshalCheckpoint() ([]byte, error)
	UnmarshalCheckpoint(blob []byte) error
	VerifyChecksum() error
}
```
Besides, the function `UnmarshalCheckpoint` is only invoked in the `GetCheckpoint` function, following the function `VerifyChecksum`. They are repeat if we verify checksum inside the function `UnmarshalCheckpoint`
```golang
// GetCheckpoint retrieves checkpoint from CheckpointStore.
func (manager *impl) GetCheckpoint(checkpointKey string, checkpoint Checkpoint) error {
	...
	err = checkpoint.UnmarshalCheckpoint(blob)
	if err == nil {
		err = checkpoint.VerifyChecksum()
	}
	return err
}
```
By the way, both cpu checkpoint and device checkpoint don't verify checksum inside function `UnmarshalCheckpoint`

Reply:
@bart0sh What do you think?

Reply:
I'm ok with the current code.